### PR TITLE
[YUNIKORN-1985] Fix possible log spew in application tryAllocate

### PR DESCRIPTION
**Pull request by brandboat**
### What is this PR for?
Introduce rate limited logger in this pr, and applied in application#tryAllocate to avoid log spew

### What type of PR is it?
* [x] - Improvement

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1985

### How should this be tested?
covered by unit test

### CODE REVIEW

File: pkg/scheduler/objects/application.go [hunk 1]
```diff
@@ -990,7 +990,7 @@ func (sa *Application) tryAllocate(headRoom *resources.Resource, preemptionDelay
 			// the iterator might not have the node we need as it could be reserved, or we have not added it yet
 			node := getNodeFn(requiredNode)
 			if node == nil {
-				log.Log(log.SchedApplication).Warn("required node is not found (could be transient)",
+				log.Log(log.SchedApplication).Debug("required node is not found (could be transient)",
```
Comment (pbacsko) [inline pkg/scheduler/objects/application.go:993]: Could you use a rate-limit approach? It's still an issue if someone decides to use Debug level.

**Review by pbacsko — CHANGES_REQUESTED** [review]
I think limiting the output to one message/sec is better. This is acceptable as a quick fix, but let's do this properly.

File: pkg/scheduler/objects/application.go (hunk 1 above)
Comment (brandboat) [reply pkg/scheduler/objects/application.go]: > Could you use a rate-limit approach? It's still an issue if someone decides to use Debug level.
> 
> I propose a natural single msg/second limit.

I've added a RateLimitedLogger in the latest commit, appreciated for the comment.

File: pkg/log/rate_limited_logger.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// RateLimitedLogger provides a logger that only logs once within a specified duration
+func RateLimitedLog(handle *LoggerHandle, every time.Duration) *rateLimitedLogger {
+	return &rateLimitedLogger{
+		logger:  Log(handle),
+		limiter: rate.NewLimiter(rate.Every(every), 1),
+	}
+}
+
+// visible for testing
+func TestingRateLimitedLog(logger *zap.Logger, every time.Duration) *rateLimitedLogger {
```
Comment (pbacsko) [inline pkg/log/rate_limited_logger.go:42]: Nit:
* rename to sth like `getTestRateLimitedLog()`
* move it to `rate_limited_logger_test.go`

Comment (brandboat) [reply pkg/log/rate_limited_logger.go]: Many thanks, addressed comments in [0376571](https://github.com/apache/yunikorn-core/pull/663/commits/0376571ce3f0d40c0f7f76cb6a9bf8bfad66f57e)

File: pkg/log/rate_limited_logger_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func TestRateLimitedLog(t *testing.T) {
+	zapLogger, logFile := createZapTestLogger()
+	logger := getTestRateLimitedLog(zapLogger, time.Minute)
+
+	// this won't last over one minute, assert there is only one record in log file
+	for i := 0; i < 10000; i++ {
+		logger.Info("YuniKorn")
+	}
```
Comment (pbacsko) [inline pkg/log/rate_limited_logger_test.go:46]: I suggest a different approach here, we don't know how much time it takes to run 10000 iterations.

For example, invoke `logger.Info("Yunikorn")` in every 100ms for 2 seconds. That means 20 calls in total. Then you check the output.

File: pkg/log/rate_limited_logger_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	logFile := fmt.Sprintf("%s/log.stdout", logDir)
+	outputPaths := []string{logFile}
+	zapConfig := zap.NewProductionConfig()
+	zapConfig.Encoding = "json"
+	zapConfig.OutputPaths = outputPaths
+
+	zapLogger, err := zapConfig.Build()
+	if err != nil {
+		panic(err)
+	}
```
Comment (pbacsko) [inline pkg/log/rate_limited_logger_test.go:81]: I did some testing, I'd opt for in memory logging. Avoid filesystem operations if possible:

```
	buf := bytes.Buffer{}
	conf := zap.NewDevelopmentConfig()
	encoderConfig := zapcore.EncoderConfig{
		MessageKey:     "msg",
		CallerKey:      "caller",
		NameKey:        "logger",
		TimeKey:        "ts",
		EncodeTime:     zapcore.EpochTimeEncoder,
		EncodeDuration: zapcore.StringDurationEncoder,
		EncodeCaller:   zapcore.ShortCallerEncoder,
	}

	wr := bufio.NewWriter(&buf)
	l := zap.New(
		zapcore.NewCore(
			zapcore.NewJSONEncoder(encoderConfig),
			zapcore.AddSync(wr),
			conf.Level,
		),
	)
	l.Info("aaaa")
	wr.Flush()   // important

```

File: pkg/log/rate_limited_logger_test.go (hunk 3 above)
Comment (brandboat) [reply pkg/log/rate_limited_logger_test.go]: copy that !

File: pkg/log/rate_limited_logger_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+	startTime := time.Now()
+	for {
+		elapsed := time.Since(startTime)
+		if elapsed > 2*time.Second {
+			break
+		}
+		logger.Info("YuniKorn")
+		time.Sleep(100 * time.Millisecond)
```
Comment (pbacsko) [inline pkg/log/rate_limited_logger_test.go:63]: Will change the timings a bit upon commit. 2 seconds is a bit too much.

Otherwise it looks good.

**Review by pbacsko — APPROVED** [review]
+1 LGTM

**Comment by pbacsko** [comment]
@brandboat could you fix the lintern issue? While at it, change the timings from 2s/100ms to 500ms/10ms so it doesn't take that much time to execute. Then I'll merge. Thanks.

**Review by pbacsko — CHANGES_REQUESTED** [review]
Lintern, timings.

**Comment by brandboat** [comment]
Already the comments in the latest commit, many thanks ! 😄

**Pull request closed without merging**